src/data_access/FirebaseRoomsDataAccessObject.java

```diff
@@ -1,27 +1,173 @@
 package data_access;
 
 import entities.auth.DisplayUser;
+import entities.auth.User;
 import entities.rooms.Message;
+import entities.rooms.Room;
+import java.io.IOException;
 import java.time.Instant;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
+import okhttp3.*;
+import org.json.JSONException;
+import org.json.JSONObject;
+import use_case.login.LoginUserDataAccessInterface;
 import use_case.rooms.RoomsDataAccessInterface;
 
 public class FirebaseRoomsDataAccessObject implements RoomsDataAccessInterface {
+
+  private final OkHttpClient client;
+
+  public FirebaseRoomsDataAccessObject(OkHttpClient client) {
+    this.client = client;
+  }
+
   @Override
-  public List<Message> getAllMessages() {
-    DisplayUser dummy_display_user = new DisplayUser("foo", "bar");
-    Instant timestamp = Instant.now();
-    Message message =
-        new Message(timestamp, "This is a test message.", dummy_display_user.getUid());
-
-    List<Message> messages = new ArrayList<>();
-    messages.add(message);
-    return messages;
+  public Room getRoomFromId(User user, LoginUserDataAccessInterface userDAO, String roomId) {
+    String idToken = userDAO.getAccessToken(user.getEmail(), user.getPassword());
+
+    String url = String.format(Constants.ROOM_URL, roomId) + "?auth=" + idToken;
+    Request request = new Request.Builder().url(url).get().build();
+
+    try {
+      Response response = client.newCall(request).execute();
+      if (response.isSuccessful()) {
+        JSONObject rooms = new JSONObject(response.body().string());
+
+        // Get messages
+        List<Message> messages = new ArrayList<>();
+        if (rooms.has("messages")) {
+          JSONObject messagesJSON = rooms.getJSONObject("messages");
+          Iterator<String> timestampIterator = messagesJSON.keys();
+          while (timestampIterator.hasNext()) {
+            String key = timestampIterator.next();
+            Instant timestamp = Instant.ofEpochSecond(Long.parseLong(key));
+            String contents = messagesJSON.getJSONObject(key).getString("contents");
+            String encodedAuthorEmail = messagesJSON.getJSONObject(key).getString("author");
+            String authorEmail = new String(Base64.getDecoder().decode(encodedAuthorEmail));
```

Comment:
This base64 decoder is showing up everywhere, can you create a facade class with documentation about why we need to base64 encode authorEmail (so that they work as firebase keys)

Reply:
Where should the Facade class be stored? In the DAO folder?